[LV][LAA] Vectorize math lib calls with mem write-only attribute

paschalis-mpeis opened the pull request:
Teach LAA to consider safe specific math lib calls which are known to
 have set the memory write-only attribute. Those attributes are set to
 calls by `inferNonMandatoryLibFuncAttrs`, in BuildLibCalls.cpp, and the
 current ones are `modf`/`modff` and `frexp`/`frexpf`.
    
This happens only when the calls are found through TLI to have vectorized counterparts.

Add tests for LAA and LoopVectorizer.

paschalis-mpeis committed:
LAA cannot vectorize lib calls like modf/modff

Functions like modf/modff are math lib calls that set memory write-only
attribute. Given that a target has vectorized mappings, LAA should allow
vectorization.

paschalis-mpeis committed:
[LV][LAA] Vectorize math lib calls with mem write-only attribute

Teach LAA to consider safe specific math lib calls which are known to
have set the memory write-only attribute. Those attributes are set to
calls by inferNonMandatoryLibFuncAttrs, in BuildLibCalls.cpp, and the
current ones are modf/modff and frexp/frexpf.

This happens only when the calls are found through TLI to have
vectorized counterparts.

paschalis-mpeis committed:
Add check for the 'memory(argmem: write)' attribute.

nikic commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2274,6 +2274,26 @@ bool LoopAccessInfo::canAnalyzeLoop() {
+  if (!F->hasFnAttribute(Attribute::AttrKind::Memory))
+    return false;
+
+  auto ME = F->getFnAttribute(Attribute::AttrKind::Memory).getMemoryEffects();
```
You can use `Call->getMemoryEffects()` instead and skip all the code above.

paschalis-mpeis replied:
Good suggestion thanks, latest commit applies this.

paschalis-mpeis committed:
Addressing reviewers

paschalis-mpeis committed:
Rebased and updated test after PR #80296

paschalis-mpeis commented:
Rebased to main after a couple of weeks of inactivity.

Note: windows x64 build failure seems unrelated; more likely a wide problem, failing at cmake configure.

nikic commented on clang/test/CodeGen/aarch64-veclib-function-calls-linear-ptrs.c:
```diff
@@ -0,0 +1,52 @@
+void modf_f32(float *in, float *out1, float *out2, int N) {
+  for (int i = 0; i < N; ++i)
+      out1[i] = modff(in[i], out2+i);
+}
```
This test needs to be in either `llvm/test/Transforms/LoopVectorize` or `llvm/test/Analysis/LoopAccessAnalysis`, depending on what exactly you want to test.

I don't really get what this test is checking for though, it doesn't look like anything was actually vectorized?

paschalis-mpeis replied:
> This test needs to be in either llvm/test/Transforms/LoopVectorize or llvm/test/Analysis/LoopAccessAnalysis, depending on what exactly you want to test.

I've used a C test, hence the placement. You'd prefer an LLVM IR test?

> I don't really get what this test is checking for though, it doesn't look like anything was actually vectorized?

Actually, the below patch removed modf/modff TLI mappings since this one was opened, due to some errors that were discovered. I'm converting this patch to a draft until those mappings are back again.
- #80296

paschalis-mpeis replied:
This was converted to LLVM tests for LoopVectorizer, and also added LoopAccessAnalysis tests.

Finally, the PR is stacked on top of #83143, which will re-introduce at least the modf/modff mappings

mgabka commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2405,6 +2421,11 @@ void LoopAccessInfo::analyzeLoop(AAResults *AA, LoopInfo *LI,
       if (I.mayWriteToMemory()) {
+        // We can safety handle math functions that have vectorized
+        // counterparts and have the memory write-only attribute set.
+        if (isMathLibCallMemWriteOnly(TLI, I))
```
can this change be tested with a loop-access analysis tests using a debug output? if yes then it does not have to depend on the TLI mappings for modf/modff

paschalis-mpeis committed:
Added LAA and LV tests

Removed C test.
Code rebased on top of patch that enables mappings for modf/modff
(among others).

paschalis-mpeis commented:
Rebased to main after #80296 was merged:
- #80296

mgabka commented on llvm/test/Analysis/LoopAccessAnalysis/attr-mem-write-only.ll:
```diff
@@ -0,0 +1,117 @@
+; RUN: opt < %s -mattr=+sve -vector-library=ArmPL -passes=inject-tli-mappings,loop-vectorize -debug-only=loop-accesses -disable-output 2>&1 | FileCheck %s
```
I think you should only be calling here : "-passes='print<access-info>' -debug-only=loop-accesses -disable-output"

paschalis-mpeis replied:
Indeed, using `-passes='print<access-info>'` is a better option here:
- it allows running LAA on the code regardless of LV's Legality result
- this would allow checking for `frexp` methods in a target agnostic manner.

Great suggestion, thanks!

mgabka commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2422,6 +2438,15 @@ void LoopAccessInfo::analyzeLoop(AAResults *AA, LoopInfo *LI,
       if (I.mayWriteToMemory()) {
+        // We can safety handle math functions that have vectorized
+        // counterparts and have the memory write-only attribute set.
+        if (isMathLibCallMemWriteOnly(TLI, I)) {
```
in the read case we also check for !VFDatabase::getMappings(*Call).empty(), but i think it should be enough that LV checks for it, that allows also to test your code in a target agnostic way, and without mappings for frexp

paschalis-mpeis commented on llvm/test/Analysis/LoopAccessAnalysis/attr-mem-write-only.ll:
```diff
@@ -0,0 +1,117 @@
+
+target triple = "aarch64-unknown-linux-gnu"
+
+; TODO: add mappings for frexp/frexpf
```
It turns out without the mappings LV Legality does not let it vectorize earlier ([here](https://github.com/llvm/llvm-project/blob/eaf0d82529f30fd53b453886676821d67ccfe9af/llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp#L900)):
> LV: Not vectorizing: Found a non-intrinsic callsite   %call = tail call float @frexpf(float noundef %0, ptr noundef %add.ptr)
> ..
>LV: Not vectorizing: Cannot prove legality

I will keep the those LAA frexp/frexpf tests, but I'll convert TODO's in a similar fashion with the [next comment](https://github.com/llvm/llvm-project/pull/78432#discussion_r1581031468).

paschalis-mpeis replied:
Short answer: Not needed.

---

**Longer answer:**
Using `-passes='print<access-info>'` (as per @mgabka suggestion), LAA pass can run regardless of vecLib existence, TLI mappings, and/or even SVE support.

Previously, in the LAA test [attr-mem-write-only.ll](https://github.com/llvm/llvm-project/pull/78432/commits/c3aa63c98627c2edf16420c5a4aa53ba233d544c#diff-6e65eaeb2e328f78d3620c102adaa22dd2a4e6bd0e677208ecb32e772c21089c) I had:
- `-mattr=+sve -vector-library=ArmPL -passes=inject-tli-mappings,loop-vectorize`

and yet, I couldn't check for the fexpr calls, as `LoopVectorizationLegality` would not allow vectorization.

With the latest patch, however, LAA runs regardless, so I'm able to do such checks already ([here](https://github.com/llvm/llvm-project/pull/78432/commits/c3aa63c98627c2edf16420c5a4aa53ba233d544c#diff-6e65eaeb2e328f78d3620c102adaa22dd2a4e6bd0e677208ecb32e772c21089cR7) and [here](https://github.com/llvm/llvm-project/pull/78432/commits/c3aa63c98627c2edf16420c5a4aa53ba233d544c#diff-6e65eaeb2e328f78d3620c102adaa22dd2a4e6bd0e677208ecb32e772c21089cR34)).

In the future, when the mappings for frexp are added, only the LV tests would need to be updated ([AArch64/veclib-function-calls-linear-ptrs.ll](https://github.com/llvm/llvm-project/pull/78432/commits/c3aa63c98627c2edf16420c5a4aa53ba233d544c#diff-7dd871d3ed08b76c541f7abb3a1d170800dab9d90dd16c4795ad747c94a17e65)).

paschalis-mpeis committed:
Addressing reviewers (2)

mgabka commented on llvm/test/Transforms/LoopVectorize/AArch64/veclib-function-calls-linear-ptrs.ll:
```diff
@@ -0,0 +1,134 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --filter "call.*(frexp|modf)" --version 4
```
sincos (which also takes linear pointers) has tests inside veclib-function-calls.ll, so I would suggest to move this tests to the same file.

paschalis-mpeis committed:
Addressing reviewers (3): tests now in veclib-function-calls.ll

Rebased history to introduce tests in 'veclib-function-calls.ll' and
delete the (now unnecessary) test veclib-function-calls-linear-ptrs.ll.

The commit history were modified as follows:
- The initial commit (that showcases what was missing) was amended to
  add the tests in veclib-function-calls.ll.
- Then, subsequent commits were similarly amended to include the updates
  that allow vectorization.
- This current commit simply dropped the no longer needed tests
  (veclib-function-calls-linear-ptrs.ll)

paschalis-mpeis commented:
### Rebased to:
1. Update to latest main
2. Amend commits for doing tests in veclib-function-calls.ll.

For (2), the commit history were modified as follows:
- The initial commit (that showcases what was missing) was amended to  add the tests in veclib-function-calls.ll.
- Then, subsequent commits were similarly amended to include the updates that allow vectorization.
- This currently latest commit simply dropped the no longer needed tests  (veclib-function-calls-linear-ptrs.ll)

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2477,6 +2493,15 @@ void LoopAccessInfo::analyzeLoop(AAResults *AA, LoopInfo *LI,
 
       // Save 'store' instructions. Abort if other instructions write to memory.
```
comment out of date

fhahn commented on llvm/test/Analysis/LoopAccessAnalysis/attr-mem-write-only.ll:
```diff
@@ -0,0 +1,116 @@
+define void @frexp_f64(ptr %in, ptr %out1, ptr %out2, i32 %N) {
+; CHECK: LAA: Allow to vectorize math function with write-only attribute: %call = tail call double @frexp
+entry:
+  %cmp4 = icmp sgt i32 %N, 0
```
this shouldn't be needed, please try to simplify the test cases

fhahn commented on llvm/test/Analysis/LoopAccessAnalysis/attr-mem-write-only.ll:
```diff
@@ -0,0 +1,116 @@
+  br i1 %cmp4, label %for.body.preheader, label %for.cond.cleanup
+
+for.body.preheader:
+  %wide.trip.count = zext nneg i32 %N to i64
```
this shouldn't be needed, please try to simplify the test cases

fhahn commented on llvm/test/Analysis/LoopAccessAnalysis/attr-mem-write-only.ll:
```diff
@@ -0,0 +1,116 @@
+  ret void
+
+for.body:
+  %indvars.iv = phi i64 [ 0, %for.body.preheader ], [ %indvars.iv.next, %for.body ]
```
please strip `%indvars.` prefix to keep value names more concise

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2477,6 +2493,15 @@ void LoopAccessInfo::analyzeLoop(AAResults *AA, LoopInfo *LI,
 
       // Save 'store' instructions. Abort if other instructions write to memory.
       if (I.mayWriteToMemory()) {
+        // We can safety handle math functions that have vectorized
```
typo: safely

fhahn commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -2477,6 +2493,15 @@ void LoopAccessInfo::analyzeLoop(AAResults *AA, LoopInfo *LI,
       // Save 'store' instructions. Abort if other instructions write to memory.
       if (I.mayWriteToMemory()) {
+        // We can safety handle math functions that have vectorized
+        // counterparts and have the memory write-only attribute set.
```
I am not sure the reasoning here is correct; the writes could still alias with other accesses,


e.g. something like below would be considered as safe, but `%out2` could overlap with `%in` in a way that results in a backwards dependence at runtime.

```
define void @frexp_f64(ptr %in, ptr %out1, ptr %out2, i32 %N) {
entry:
  %cmp4 = icmp sgt i32 %N, 0
  br i1 %cmp4, label %for.body.preheader, label %for.cond.cleanup

for.body.preheader:
  %wide.trip.count = zext nneg i32 %N to i64
  br label %for.body

for.cond.cleanup:
  ret void

for.body:
  %indvars.iv = phi i64 [ 0, %for.body.preheader ], [ %indvars.iv.next, %for.body ]
  %arrayidx = getelementptr inbounds double, ptr %in, i64 %indvars.iv
  %0 = load double, ptr %arrayidx, align 8
  %add.ptr = getelementptr inbounds i32, ptr %out2, i64 %indvars.iv
  %call = tail call double @frexp(double noundef %0, ptr noundef %add.ptr)
  store double %call, ptr %arrayidx, align 8
  %indvars.iv.next = add nuw nsw i64 %indvars.iv, 1
  %exitcond.not = icmp eq i64 %indvars.iv.next, %wide.trip.count
  br i1 %exitcond.not, label %for.cond.cleanup, label %for.body
}

declare double @frexp(double, ptr) #1
attributes #1 = { memory(argmem: write) }
```

mgabka replied:
Hi @fhahn,
you are right.

The whole logic was based on checking later in the  https://github.com/llvm/llvm-project/blob/332de4b2677ce7a95cc2df30d761fbb55376fe07/llvm/lib/Transforms/Vectorize/LoopVectorize.cpp#L6659 if the pointers are linear with the right step, but there is no checks for aliasing.
@paschalis-mpeis that means that there is an existing bug with sincos, as following code gets vectorised, without any checks:

```
double sin[N];
double cos[N];
sin[5] = M_PI;

for(int i=0; i<N; i++ ) {
 sincos(sin[5], sin+i, cos+i);
}

```
